### ISSUE TITLE
Discussion

### ISSUE THREAD

**Issue by marcpope** [marcpope/borgbackupserver#1]
Feel free to discuss Borg Backup Server here.

**Comment by cganiddle** [comment]
What is the status of this interesting project ?
Is there a version available to do tests?
Will it be open-source? 

Thanks for the feedback

**Comment by swissbyte** [comment]
Hi Marc

Thanks for your effort on this project in advance. 
There is really a need for such a solution out there (myself included). 
The question about open source is understandable. Im also a software engineer and know what it means to upload hundreds of hours of code "for free". But i think you could also get a lot back from the community if you make it open source. There will be definitley many contributors out there. 
On the other hand, i would suggest to use patreon or other platforms to earn some money. 
You could keeps some features closed for supporters and release them 6 month later to the community as soon as they are "paid back" from the supporters which usually are companys. 
That way the code is available for everyone but the ones who need new features earlies, can pay for them. 

Take a look at sloeber (arduino eclipse IDE) jantje does practice this model of payment. And it looks like that it is a win win situation for everyone. 

I would also pay some bucks for your solution if possible. 
Since im a stundent i cant pay thousands of dollars. but i would support you with a few dollars per month. 

By the way, whats the status of the projects?
Regards

**Comment by Marekxbx** [comment]
I agree with swissbyte.
Borgbackup is a great tool butwith your gui and integration will be sensational. 
I will support financially (probably not only me) your work on the project.

**Comment by marcpope** [comment]
This project has been almost 2 years in the making. 

The current rationale: the project will be open-source and I will accept bounties for features and paid support options such as installation and maintenance. We will also offer a hosted backup solution if you don't want to run your own server.

I wanted to start building Borg Backup Server because I was tired of R1Soft's crazy business practices and ridiculous pricing structures and I really liked the way borg works.  I need to read up more on open-source models and how to protect certain aspects of the software, such as how we encrypt and store the Repository Keys. Security is important. If a hacker gets a hold of a client machine, we don't want them hurting the backups. 

This product is designed from an ISP perspective. I have 100s of public facing web servers that I backup on a daily and even hourly basis. 

If there is any money made, at least 25% will go back to the open-source community project of borg backup. I am not building this project to get rich. I don't need it to support myself. It's important to me that the open-source borg project stays alive and grows. I'd like to see a Windows client that is official. I'd also like to see borg allow reads during a write lock.

I am afraid I am still a long ways away from having a beta that can be tested. I'm not going to release something that is half-baked. It's going to be as solid as possible out-of-the-gate. Backups are important, and it's critical they work.

I understand a lot of people are interested in the project and it is a priority to get it done as soon as possible, but you just have to be patient. If you can't wait, there are some other projects out there that may suit your needs for now.

I have big goals for the project! It will be worth the wait. I plan to release some more videos and screenshots in the near future.

**Comment by colorcube** [comment]
This looks very nice and I hope it will be released soon!

I have two questions:

1. Will the clients be pulled? 
That way it would be possible to have a Backup Server which is not accessible from the internet.

2. This is clearly designed to make backups from a bunch of servers, but will it be possible to integrate a laptop (which is not always connected) in such an environment?
... without getting hundreds of warning emails

**Comment by marcpope** [comment]
1. The server connects to the clients over ssh. So as long as it can connect to them it will work.

2. The client machines will need a constant hostname or IP address. Also this will only support Linux and maybe Mac clients as there is no official Borg for Windows. As far as warnings I could add something that says don’t warn to the settings?

**Comment by colorcube** [comment]
Sounds good

The laptop is running Linux ;-)
An option to disable warnings might be nice for machines that aren't always online.
Later that could be extended to warn when x days the backup failed.

**Comment by marcpope** [comment]
You can also do on-demand from the server. Like "Backup Now". But yes I think that would be a good idea for machines that are not always on the network.  I have toyed with several ideas on how to handle remote machines that are behind firewalls. Right now, I am testing using autossh to maintain a tunnel to Borg Backup Server so that it can connect and issue commands. It works fairly well.  But another thing that might work for laptops, is just a cron job that the server generates that tells the clients to check in every so often and the server would issue commands. This way, they wouldn't need to have a static IP or hostname.  I am open to other ideas.. this probably won't make it into the initial version, but it's something we have a need for.

**Comment by marcpope** [comment]
As I'm building the software, I'd thought I would give a sneak peak of some features. The software now has the ability to backup clients that have non-static IPs or clients that are behind firewalls using a Reverse SSH Tunnel. The client will be instructed to run a script that will automatically setup the tunnel using AutoSSH and install the proper Init Scripts. Here's a quick screenshot:

<img width="1315" alt="new-client" src="https://user-images.githubusercontent.com/11093841/46054045-6baf7380-c113-11e8-9ab0-a3365857378c.png">

**Comment by marcpope** [comment]
This is the latest dashboard.. the notifications at the top are still a work in progress.
<img width="1317" alt="dashboard" src="https://user-images.githubusercontent.com/11093841/46054566-da8dcc00-c115-11e8-80ed-2e77f7b38956.png">

**Comment by poxin13** [comment]
Just came upon this repo when setting up Borg for the first time on my home server. Funny enough, I also work for a small web hosting company utilizing r1soft. Yeah it can get expensive pretty quick, though great for cPanel boxes.

Any plans on a scheduled release for this? Would love to give it a go on my home setup.

**Comment by marcpope** [comment]
Since I’m doing it in my spare time and my other programmer is ill, it’s not going fast but I am working on it every day. I hope by the end of this year.

**Comment by poxin13** [comment]
Sounds good, will keep an eye out. Cheers from NC as well 👍

**Comment by Ulli2k** [comment]
I am really looking foreward to your first release.
I would appreciate an early version with reduced functionality..you could add functions step by step and we will test it?

**Comment by marcpope** [comment]
I've been working hard on Borg Backup Server over the past month and a lot of progress has been made. The interface is now even cleaner and many bugs have been squished. We are working on the Queue and the Recovery Options sections this week. Here's a sneak peak of the schedule page. Hope to have a productive week ahead.

![schedulepage](https://user-images.githubusercontent.com/11093841/47954156-51ac5080-df5d-11e8-91f9-737476eae3e2.png)

**Comment by Marekxbx** [comment]
You are the hero Mark.  We are waiting for the first edition.

**Comment by McLive** [comment]
Indeed. Looks awesome!

**Comment by Ulli2k** [comment]
Wow awesome!

**Comment by tentious** [comment]
Looks great! Can't wait!

**Comment by marcpope** [comment]
Detailed job status is now a reality. This is something I really wanted. Took several days and a trip to the chiropractor for a stiff neck!

![ttulor](https://user-images.githubusercontent.com/11093841/48321014-3391c080-e5ed-11e8-8d53-ec75912667d6.jpg)

**Comment by Ulli2k** [comment]
Nice  feature!

**Comment by Ulli2k** [comment]
Looks like the basic backup function is already implemented?
great features are currently following?
Could that be a point were we could try an early release?

**Comment by Azertooth** [comment]
looks awesome! keep up the good work and thank you for it!

**Comment by rmanibus** [comment]
Awesome project! There is no accessible backup solution today on the market.
Do you need help for the development ?

**Comment by ang3lx** [comment]
Any update?

**Comment by marcpope** [comment]
Not yet, we’ve been working hard. Hopefully reporting some new stuff soon!

**Comment by bilyboy785** [comment]
Hello Team ! 
Any news ? :)

**Comment by andererbenutzer** [comment]
Indeed, it would be awesome to get some update on the project.
It appears to be an amazing tool.

**Comment by McLive** [comment]
Yeah, can't wait to see this released 😨

**Comment by marcpope** [comment]
This project is very much alive as being worked on. Due to some flu and family matters, we had to take a couple months off development but we are actively working on it again mostly on the internal engine, and better error recovery and reporting.

Hope to show you guys more info soon.

Marc

**Comment by Vladikamira** [comment]
Get well soon, Mark! Waiting for the first release of it! :)

**Comment by gamerton** [comment]
Waiting for release) Looks great!

**Comment by bleed2002** [comment]
I recently switched from duplicity to borg backup, it's awesome. But a management tool like this is really missing, so I appreciate your work on this and i would be happy to help by testing the first release ;-) Thanks

**Comment by marcpope** [comment]
Hi Patrick:

 

Thanks for the feedback and I know there’s a lot of people interested in this project. We are very committed to getting it done as soon as possible as it would save us a lot of money on our R1Soft licenses we currently are paying and give us greater controls over our backups. The goal right now is to get it to a testable milestone so we can start getting some feedback. 

 

I do have a question to ask the community:

 

Do you feel SSH is an acceptable way for the server to communicate with the backup clients? Or would you rather have a Daemon/Service running on each client and the server communicate through that. I feel a Daemon/Service would have some extra security benefits, but I am unfamiliar with that sort of programming, but not against farming that part out.

 

I know some people have security restrictions that prevent using ssh, but since borg’s primary method of backup is ssh keys, it seems the most logical.

 

Thanks

Marc

**Comment by ronindesign** [comment]
My perspective on this, is the issue / choice is not weighted enough to be worth delaying the project.

If you can out source this while you work on the core, then great!
Otherwise, I'd say delaying this to roll your own "protocol" when SSH is plenty sufficient is not worth.
People have lots of experience getting it to work in environments where it's not technically supported out-of-box (e.g. tunnels, proxies, etc).
Additionally, some might prefer the well-tested and understood transport via SSH rather than a custom roll.

So, bottom line, unless this adds (or allows for possibility of) amazing features / capabilities, SSH works just fine and there's no need to suffer the overhead of adding in extra sub-system for managing communication.

**Comment by pille** [comment]
@marcpope 
> I do have a question to ask the community: Do you feel SSH is an acceptable way for the server to communicate with the backup clients? 

SSH is perfectly fine for me. it can have its own restricted user inside a jail and i prefer well-known protocols.

**Comment by colorcube** [comment]
I fully agree with @ronindesign

**Comment by robaato** [comment]
I agree with @ronindesign too.

**Comment by McLive** [comment]
I agree with @ronindesign too.

**Comment by jfguyet** [comment]
With my Agile mindset, I would tend to say, MVP version (Minimum Viable Version) would run with SSH. It would do the job for most users. Eventually, if the need is confirmed, I would roll out my own protocol.
Also don't forget designing and building a secure protocol is a very complex task. I mean almost (if not all) protocols have or had vulnerabilities, right ?

**Comment by bleed2002** [comment]
I also completely agree with @ronindesign, ssh is absolutely fine and even prefered for me now.

**Comment by Ulli2k** [comment]
I would definitly prefer ssh. 
Because 
* every client already use it and it is maintained and security issues are getting closed by a big community.
* it safes a lot of work. ;)
* state of the art

**Comment by StopTrain** [comment]
Agreed, ssh is just fine for me as well.   really looking forward to this... looks amazing!

**Comment by airbillion** [comment]
Hello,

Just wanted to say that this project sounds great!!

I have been using Borg for the past few years and a web interface would be amazing!

Any ideas when a beta version will be available for testing?

Thank you!

**Comment by MichaelTrip1986** [comment]
Hi Marc,

I will be watching this project closely. Can't wait to try your first release!

**Comment by NeoMatrixJR** [comment]
I agree with the Agile practices.  Minimum Viable Product!!! Stop adding any nice-to-have features if you're still coming up with "well it'd be great if...." ideas.  Write them down, but don't do anything else until your existing featureset is done.  Back off any non-core/non-necissary features until core product is working.  Even if the product isn't done, you can release it/show it off.  The more people/users you have looking at something early on the more time you can save by fixing things or changing course early.  If your product is at a workable level...even if some things have to be done manually, I'd say get it out there...at least to some test users!

**Comment by marcpope** [comment]
We are doing what I consider minimal at this stage. There’s many features also planned for the future. Yes,  it is taking a long time.  However, when you are dealing with backups, you need something rock-solid. Borg is great, but making an interface for it is not the easiest, especially with it’s lack of simultaneous backup/restore and directory listing. We’ve had to build a task manager to be the traffic cop, we’ve had to build an indexer to be able to select files to restore while not interrupting backups, etc. There’s a lot to it and it’s my guess why nobody has built anything substantial yet.

 

I may have jumped the gun showing it off because people are fairly impatient these days. If you need something right away, I suggest you look at other solutions. This may not be for you.

 

At our company, we use R1Soft. I am paying $800 a month for licenses. It’s insane. So I have a major motivator to get it done so I don’t have to pay that any longer. It’s the prime reason I decided to make something because there is nothing affordable out there that does what I want. In the end, I think it will be worth it.

 

-Marc

**Comment by TuRDMaN** [comment]
> My perspective on this, is the issue / choice is not weighted enough to be worth delaying the project.
> 
> If you can out source this while you work on the core, then great!
> Otherwise, I'd say delaying this to roll your own "protocol" when SSH is plenty sufficient is not worth.

Just wanted to say that I completely agree with this. SSH is a must-have, and your daemon/service implementation could possibly be a nice addition further down the road, after release.

Thanks for all your hard work. For now, I'll be waiting patiently until we get a release to play around with!

**Comment by marcpope** [comment]
The dashboard has taken shape. Every aspect of the stats update in real time using AJAX/JQuery. It's designed so you can leave it open on a monitor and keep an eye on your backups. Keep in mind, if you are someone like me, we are backing up 100s of physical and virtual servers. Backups are essential to our company. 

You can see the dashboard has changed quite a bit since we've first showed it off. Now it's using real data from active backups. I am very excited to get a beta out as soon as I can. I've been working on it 12-14 hours a day lately.

![Borg-Dashboard](https://user-images.githubusercontent.com/11093841/55446379-76990400-558d-11e9-8f2d-d8f568ec8534.png)

**Comment by bleed2002** [comment]
Thanks for the effort you put into this, cant wait to try it out.

**Comment by Azertooth** [comment]
It's coming up so beautiful, thank you for your effort. Regarding the daemon/ssh, would be indeed a nice addition to the project but we can have them in a second time without further delaying the first beta release


________________________________
Da: Patrick Kaiser <notifications@github.com>
Inviato: mercoledì 3 aprile 2019 10:22
A: marcpope/borgbackupserver
Cc: Azertooth; Comment
Oggetto: Re: [marcpope/borgbackupserver] Discussion (#1)

Thanks for the effort you put into this, cant wait to try it out.

**Comment by marcpope** [comment]
It uses SSH already, that’s what I built it with and that’s what we will be going with. It makes the most sense since borg itself uses SSH too.

 

-Marc

**Comment by Ulli2k** [comment]
Looks awesome and very professional!

**Comment by feldsam** [comment]
Hello @marcpope, thank you for great work! Let me introduce our company FeldHost™. We are hosting and development company based in Prague - Czech republic, we use and support OpenNebula Cloud management platform. We already develop backup solution for OpenNebula Virtual Machines and it is published as official addon there https://github.com/OpenNebula/addon-image-backup. Few weeks ago, I discovered Borg Backup and today I found your Borg Backup Server GUI. We like to participate on development of Borg Backup Server GUI. If you like have more developer hands on, them I will be delighted. Let me know if you are interested, you can find my contact details on my profile. Thank you!

**Comment by d2dyno** [comment]
Just found this while looking for a central UI for my server backups. I look forward to it!

**Comment by isclever** [comment]
Staring this and hope you release this great looking tool. 

A nice feature (which you might already have) is a way of syncing the repository (or a configurable subnet) to S3/B2/anything for off site backups.

**Comment by marcpope** [comment]
> Staring this and hope you release this great looking tool.
> 
> A nice feature (which you might already have) is a way of syncing the repository (or a configurable subnet) to S3/B2/anything for off site backups.

That is something we are looking at. Won't be in the first release but definitely on my mind. We are still a little ways off from getting this done. We are working on the restore section now. Once that is in place, we hope to release a beta version for people to try and give us feedback.

**Comment by Ulli2k** [comment]
Any news on the first beta release?
sorry I not not want to bother you but I am really looking forward to it ;)

**Comment by ronindesign** [comment]
I'm still looking forward to this as well, especially with the v1 release of Duplicacy's Web UI:
https://forum.duplicacy.com/t/duplicacy-web-edition-1-0-0-is-now-available/2053/9

I'm still waiting to make my final decision between Borg and Duplicacy.
I'd much prefer Borg as the Duplicacy bus number is basically 1...

Thanks for your hard work on this and am excited to see what results.

**Comment by marcpope** [comment]
There's still lots of work to do. I know people are anxious. I was hesitant to even talk about it because I know people would be antsy. This has been two years in the making and there's still a lot to do. I know people want to try it out, but I can't release something that has bugs since we are dealing with people's data.

I am not familiar with Duplicacy's Web UI. Is that just for controlling a single client or a lot of clients? Borg Backup Server is centralized backup server that controls clients and keeps track of schedules, file searching/indexing, restores and more. It will make it easy to monitor all your backups. It is designed to backup lots of clients and might be overkill for anything less than a few. Currently, I have the software backing up over 75 machines and it has been running for weeks with no issues.

Our indexing, search and restore features are in the works. Once those are completed, I plan to release a beta version. I can't tell you exactly when it will be, but I hope in the next couple months.

**Comment by Ulli2k** [comment]
Would it be possible to release the backup functionality without the indexing, search and restore feature? Sorry for bothering you ;)

**Comment by l0nax** [comment]
@marcpope Whats the status of the Project?

**Comment by McLive** [comment]
> @marcpope Whats the status of the Project?

You might read the messages above. ;)

**Comment by bend** [comment]
Very interesting project. Would love to help on that once it is released ! Do you have an ETA to release a beta ?

**Comment by megusta-01** [comment]
Nice project, clean design, thumbs up! Very usefull for noobs like me. 
It will be great if you can make it for free for private users, maybe only for 5 clients or something like that

**Comment by Ulli2k** [comment]
Hey Marc,
how is it going, did you make a good progress during the last weeks?

**Comment by Paulikid** [comment]
> Nice project, clean design, thumbs up! Very usefull for noobs like me.
> It will be great if you can make it for free for private users, maybe only for 5 clients or something like that

I agree, very nice project for backing up home servers, too.

**Comment by Tomw787** [comment]
Hey, hows it going? 

We haven't heard from you for awhile!

**Comment by Tom-Development** [comment]
Hey, hows it going?

We haven't heard from you in awhile!

**Comment by marcpope** [comment]
hey guys. The wife was pregnant and had a bunch of complications which was eating up a lot of my time. We just had a beautiful baby girl at 4.5lb! I hope to get back to work in the next week or so. Things have been moving along. The software is working mostly, still working on the restore functions and a few other bugs.

-Marc

**Comment by Azertooth** [comment]
congratz mate! <3 have a good time with your family 


________________________________
Da: Marc Pope <notifications@github.com>
Inviato: martedì 13 agosto 2019 13:37
A: marcpope/borgbackupserver
Cc: Azertooth; Comment
Oggetto: Re: [marcpope/borgbackupserver] Discussion (#1)

hey guys. The wife was pregnant and had a bunch of complications which was eating up a lot of my time. We just had a beautiful baby girl at 4.5lb! I hope to get back to work in the next week or so. Things have been moving along. The software is working mostly, still working on the restore functions and a few other bugs.

-Marc

**Comment by maiksd** [comment]
Hey Marc, congrats on the baby, take your time!

I just wanted to chime in that I'm also in the queue waiting for a release. Currently we're using a custom script which runs on a dedicated backup server, using rsync to collect backups from the clients and storing them on NFS storage. We would like to switch to borg to facilitate offsite backups.

So we're already relying on SSH to connect to the clients, and I would like to leave it that way because not having to install/maintain any software on clients is a major benefit. Just put an authorized_keys file there and done.

Will borgbackupserver support multiple backup targets, with different schedules?

**Comment by kramer2** [comment]
update please ?

**Comment by megusta-01** [comment]
with a baby there is no time for hobbies, so bad luck for you guys :)

**Comment by marcpope** [comment]
Sorry for the delayed response. 

Its still being worked on, we’ve found some bugs that have been causing errors we are trying to work out and we need more error handling to see what’s happening. I have been fairly busy and not had tons of time to work on it but I’ve spent at least 15-20 hours in the last month on it. 

My other programmer has been out a lot lately with the flu so that’s also caused delays. It’s definitely something we want to turn out but it’s not in a place for that yet.

-Marc

**Comment by T-6** [comment]
Would be really nice if this work would be published.

**Comment by feldsam** [comment]
@marcpope, I think when published, many contributions can arrive, so in the end you save time and work.

**Comment by janszczepanski** [comment]
Any plans when you'll release it? I general I think you could make a business out of it, so many people are interessted since years ;) Looking forward to be able to use it someday (maybe 2019?)

**Comment by ryszard-suchocki** [comment]
Have you seen [CyBorgBackup] (https://github.com/gaetanf/cyborgbackup) ? It is very similar to Marc Pope's Borg backup server.

**Comment by d2dyno** [comment]
I hadn't, looks like a great project for the time being, thanks @ryszard-suchocki

**Comment by insanerwayner** [comment]
Definitely open this project up. I think you'll make a lot more traction with a community involved in the development.

**Comment by sapientsaxonsaboo** [comment]
Hi @marcpope I use Borg myself to backup various Linux clients and servers. Was looking for a web-based GUI and a way to more easily deploy and manage Borg centrally. If not, then to create something. Came across your project quite a while ago and it looks just the ticket. Have been watching since. 

I've read your comments on open-sourcing the project. Any further thoughts on that at this juncture? Maybe the community could support and contribute. I'm sure there are many like-minded. 

-- SSS

**Comment by marcpope** [comment]
I understand everyone's desire to get this, and I apologize it's taking longer than expected due to the complexity of many different setups, error handling, etc. We are working on it as fast as possible. The project will be available for free for a limited number of servers and it will be commercial for larger installations. A portion of sales will also be donated to fund ongoing open-source borg development.  Unfortunately, we are not in a position to open-source the project as there is just too much time and development money spent on it. I do not have a pricing structure at this time but it will be very affordable and significantly less expensive than other commercial backup systems. We have it being used in about a dozen locations testing and it's doing a great job backing up.

**Comment by sapientsaxonsaboo** [comment]
Hi Marc, thanks for the response. That's a shame you're choosing not to open source it when it's based on an existing open source tool. 

I hope the tool is a commercial success for you, and I'm sure the contribution to Borg will be welcome. 

-- SSS

**Comment by marcpope** [comment]
There’s 1000s of commercial tools based on open source projects.

**Comment by megusta-01** [comment]
Don't disappoint and angry about some people here, i can understand that donation sadly isn't enought to maintain projects. Your decision is reasonable compromise that have to accept. 
Keep in mind opensource has also his benefits especially that community can help and maintain this project.
Good luck

**Comment by bb2k** [comment]
no opensource ... see ya then .....

**Comment by kelson42** [comment]
@marcpope Do you have a first release timeline? This will be a software release? with Docker? Or do we talk here about SAS?

**Comment by marcpope** [comment]
Sorry no timeline yet. This will be a software release. Considering Docker. I’ve not used it before so I’m not sure how involved it will be. There will be an installer. It will want a clean system. It will only be a CentOS to start as the server but am looking at Ubuntu.

I am working on making it a little more flexible for different environments and pushing myself to get a beta out. 

It’s been running at my web hosting company now for 8 months with very few issues and it backs up 150 servers without any problem.

Marc

**Comment by bilyboy785** [comment]
Each week; new story with new date but nothing online ! Bullshit

**Comment by marcpope** [comment]
You’re welcome to make your own
